src/main/java/com/playmonumenta/networkrelay/RemotePlayerAbstraction.java

```diff
@@ -0,0 +1,155 @@
+package com.playmonumenta.networkrelay;
+
+import com.google.gson.JsonObject;
+import com.google.gson.JsonPrimitive;
+import com.playmonumenta.networkrelay.util.MMLog;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.UUID;
+import org.jetbrains.annotations.Nullable;
+
+public abstract class RemotePlayerAbstraction {
+	/**
+	 * The ID of the server ID reporting this information for a given server type.
+	 * For example "proxy" could report "bungee-13", or "minecraft" could report "valley-2".
+	 * This is not used for proxies to report which Minecraft instance is most relevant.
+	 */
+	protected final String mServerId;
+	/** The player's Minecraft UUID */
+	protected final UUID mUuid;
+	/** The player's name */
+	protected final String mName;
+	/** Whether the player is online; this is broadcast as offline to remove remote players from local caches */
+	protected final boolean mIsOnline;
+	/** Whether the player is visible to most players or not. Is null if unset */
+	protected final @Nullable Boolean mIsHidden;
+	/** Data provided by other plugins */
+	protected final Map<String, JsonObject> mPluginData;
+
+	/**
+	 * Data from this (local) server
+	 *
+	 * @param serverId - server id (shard name)
+	 * @param uuid - player's uuid
+	 * @param name - player's name
+	 * @param isOnline - player's online status
+	 * @param isHidden - player's visibility status
+	 */
+	protected RemotePlayerAbstraction(String serverId, UUID uuid, String name, boolean isOnline, @Nullable Boolean isHidden) {
+		mServerId = serverId;
+		mUuid = uuid;
+		mName = name;
+		mIsOnline = isOnline;
+		mIsHidden = isHidden;
+		mPluginData = new HashMap<>();
+	}
+
+	/** Data from a remote server of any type (subclasses provide extra information) */
+	public RemotePlayerAbstraction(JsonObject remoteData) {
+		mServerId = remoteData.get("serverId").getAsString();
+		mUuid = UUID.fromString(remoteData.get("playerUuid").getAsString());
+		mName = remoteData.get("playerName").getAsString();
+		mIsOnline = remoteData.get("isOnline").getAsBoolean();
+		JsonPrimitive isHiddenJson = remoteData.getAsJsonPrimitive("isHidden");
+		if (isHiddenJson == null || !isHiddenJson.isBoolean()) {
+			mIsHidden = null;
+		} else {
+			mIsHidden = isHiddenJson.getAsBoolean();
+		}
+		mPluginData = new HashMap<>();
+		JsonObject pluginData = remoteData.getAsJsonObject("pluginData");
+		for (String key : pluginData.keySet()) {
+			mPluginData.put(key, pluginData.getAsJsonObject(key));
+		}
+	}
+
+	/** Determine the appropriate remote player data type to use */
+	public static @Nullable RemotePlayerAbstraction from(JsonObject remoteData) {
+		if (remoteData == null) {
```

Comment:
remoteData is not nullable - this should never be allowed to happen, which means this method does not need to return nullable either.